### PR TITLE
[CPU] Fix mixing VEX and non-VEX instructions

### PR DESCRIPTION
**Pull request by EgorDuplensky**
-------

### Details:
 - Quote: The Skylake microarchitecture implements a different state
 machine than prior generations to manage the YMM state transition
 associated with mixing SSE and AVX instructions.
 It no longer saves the entire upper YMM state when executing
 an SSE instruction when in “Modified and Unsaved” state,
 but saves the upper bits of individual register.
 As a result, mixing SSE and AVX instructions will experience
 a penalty associated with partial register dependency of
 the destination registers being used and additional blend
 operation on the upper bits of the destination registers.

    Such type of penalties have a huge impact on openvino's and oneDNN's kernels.
 Basically the mixing of VEX and non-VEX instructions should be
 avoided.

- The PR addresses only known issues with mixing vex / nonVEX.
  Preventing this issue for all the cases and automatic verification will be done in scope of separate activity

- This is basically mapping of https://github.com/openvinotoolkit/openvino/pull/7238 to the master branch
with very few fixes.

### Tickets:
 - 65245
### oneDNN PR:
- https://github.com/openvinotoolkit/oneDNN/pull/94

### CODE REVIEW

**Comment by dmitry-gorokhov** [comment]
@chenhu-wang please review the PR

**Comment by chenhu-wang** [comment]
for https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/emitters/jit_eltwise_emitters.cpp#L650
As uni_vcmpeqps() and uni_vblendps() should be very carefully to create(cmpps have different imm, 8 for non-vex and 32 for vex, fixed xmm0 for blendps() on non-vex),
how about change to:
```
if (mayiuse(avx512)) {
        h->vcmpps(k_mask, vmm_src0, vmm_src1, _cmp_eq_oq);
        h->uni_vmovups(vmm_dst, table_val("zero"));
        h->vblendmps(vmm_dst | k_mask, vmm_dst, table_val("one"));
} else if (mayiuse(cpu::x64::**avx**)) {
        h->vcmpeqps(vmm_aux0, vmm_src0, vmm_src1);
        h->uni_vmovups(vmm_dst, table_val("zero"));
        h->uni_vblendvps(vmm_dst, vmm_dst, table_val("one"), vmm_aux0);
} else {
        h->movups(vmm_aux0, vmm_src0);
        h->cmpps(vmm_aux0, vmm_src1, _cmp_eq_oq);
        h->movups(vmm_aux1, table_val("one"));
        h->pxor(vmm_dst, vmm_dst);
        h->blendvps(vmm_dst, vmm_aux1);
}
```
Then avx target will run vex instructions with xmm registers.

**Comment by chenhu-wang** [comment]
> for
> 
> https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/emitters/jit_eltwise_emitters.cpp#L650
> 
> 
> As uni_vcmpeqps() and uni_vblendps() should be very carefully to create(cmpps have different imm, 8 for non-vex and 32 for vex, fixed xmm0 for blendps() on non-vex),
> how about change to:
> ```
> if (mayiuse(avx512)) {
>         h->vcmpps(k_mask, vmm_src0, vmm_src1, _cmp_eq_oq);
>         h->uni_vmovups(vmm_dst, table_val("zero"));
>         h->vblendmps(vmm_dst | k_mask, vmm_dst, table_val("one"));
> } else if (mayiuse(cpu::x64::**avx**)) {
>         h->vcmpeqps(vmm_aux0, vmm_src0, vmm_src1);
>         h->uni_vmovups(vmm_dst, table_val("zero"));
>         h->uni_vblendvps(vmm_dst, vmm_dst, table_val("one"), vmm_aux0);
> } else {
>         h->movups(vmm_aux0, vmm_src0);
>         h->cmpps(vmm_aux0, vmm_src1, _cmp_eq_oq);
>         h->movups(vmm_aux1, table_val("one"));
>         h->pxor(vmm_dst, vmm_dst);
>         h->blendvps(vmm_dst, vmm_aux1);
> }
> ```
> 
> replace mayiuse(**avx2**) with mayiuse(**avx**). then avx target will run vex instructions with xmm registers. The same for all eltwise node of compare.

or sth like this with less mayiuse() called:
```
if (isa == cpu::x64::avx512_common) {
        ...
} else if (isa == cpu::x64::avx2 || mayiuse(avx)) {
        ...
} else {
        ...
}
```

**Comment by EgorDuplensky** [comment]
> > for
> > https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/emitters/jit_eltwise_emitters.cpp#L650
> > 
> > As uni_vcmpeqps() and uni_vblendps() should be very carefully to create(cmpps have different imm, 8 for non-vex and 32 for vex, fixed xmm0 for blendps() on non-vex),
> > how about change to:
> > ```
> > if (mayiuse(avx512)) {
> >         h->vcmpps(k_mask, vmm_src0, vmm_src1, _cmp_eq_oq);
> >         h->uni_vmovups(vmm_dst, table_val("zero"));
> >         h->vblendmps(vmm_dst | k_mask, vmm_dst, table_val("one"));
> > } else if (mayiuse(cpu::x64::**avx**)) {
> >         h->vcmpeqps(vmm_aux0, vmm_src0, vmm_src1);
> >         h->uni_vmovups(vmm_dst, table_val("zero"));
> >         h->uni_vblendvps(vmm_dst, vmm_dst, table_val("one"), vmm_aux0);
> > } else {
> >         h->movups(vmm_aux0, vmm_src0);
> >         h->cmpps(vmm_aux0, vmm_src1, _cmp_eq_oq);
> >         h->movups(vmm_aux1, table_val("one"));
> >         h->pxor(vmm_dst, vmm_dst);
> >         h->blendvps(vmm_dst, vmm_aux1);
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > replace mayiuse(**avx2**) with mayiuse(**avx**). then avx target will run vex instructions with xmm registers. The same for all eltwise node of compare.
> 
> or sth like this with less mayiuse() called:
> 
> ```
> if (isa == cpu::x64::avx512_common) {
>         ...
> } else if (isa == cpu::x64::avx2 || mayiuse(avx)) {
>         ...
> } else {
>         ...
> }
> ```

I agree that this would be better and more properly structured way to handle this,
but this will potentially affect AVX (not AVX2) architectures if I understand this correctly.
Not sure if we should "touch" this in scope of the current PR.

**Comment by chenhu-wang** [comment]
some more non-vex need update. Some is in common path, some for in sse41 path. I am not sure we need update for AVX machine(isa == sse41 code path) in this PR, for sure common path have higher priority. Maybe @dmitry-gorokhov could suggest.
https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/nodes/mkldnn_bin_conv_node.cpp#L255
https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/nodes/mkldnn_bin_conv_node.cpp#L270
https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/nodes/mkldnn_bin_conv_node.cpp#L276
https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/nodes/mkldnn_bin_conv_node.cpp#L289
https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/nodes/mkldnn_bin_conv_node.cpp#L295
https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/nodes/mkldnn_bin_conv_node.cpp#L352-L353
https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/nodes/mkldnn_bin_conv_node.cpp#L361-L366
https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/nodes/mkldnn_bin_conv_node.cpp#L718

**Comment by chenhu-wang** [comment]
https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/nodes/mkldnn_eltwise_node.cpp#L764
https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/nodes/mkldnn_eltwise_node.cpp#L769
https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/nodes/mkldnn_eltwise_node.cpp#L775
https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/nodes/mkldnn_eltwise_node.cpp#L781

**Comment by chenhu-wang** [comment]
https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/nodes/mkldnn_fake_quantize_node.cpp#L774
`inline void store_vector(const Xbyak::Address &op, Xmm xmm_dst, Precision dst_prc) {}`
although this is xmm version store_vector, we still need update to uni_*, as it is used as tail4block process on skylake
https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/nodes/mkldnn_fake_quantize_node.cpp#L565

so like this need update:
https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/nodes/mkldnn_fake_quantize_node.cpp#L787
https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/nodes/mkldnn_fake_quantize_node.cpp#L792

**Comment by dmitry-gorokhov** [comment]
> some more non-vex need update. Some is in common path, some for in sse41 path. I am not sure we need update for AVX machine(isa == sse41 code path) in this PR, for sure common path have higher priority. Maybe @dmitry-gorokhov could suggest.

Lets not include AVX path improvements in this PR. It should be separate feature. Please create follow up ticket

**Comment by dmitry-gorokhov** [comment]
> https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/nodes/mkldnn_fake_quantize_node.cpp#L774
> 
> 
> `inline void store_vector(const Xbyak::Address &op, Xmm xmm_dst, Precision dst_prc) {}`
> although this is xmm version store_vector, we still need update to uni_*, as it is used as tail4block process on skylake
> https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/nodes/mkldnn_fake_quantize_node.cpp#L565
> 
> so like this need update:
> 
> https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/nodes/mkldnn_fake_quantize_node.cpp#L787
> 
> 
> https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/nodes/mkldnn_fake_quantize_node.cpp#L792

@EgorDuplensky Seems like FQ should be fixed in bounds of this PR

**Comment by EgorDuplensky** [comment]
> > https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/nodes/mkldnn_fake_quantize_node.cpp#L774
> > 
> > `inline void store_vector(const Xbyak::Address &op, Xmm xmm_dst, Precision dst_prc) {}`
> > although this is xmm version store_vector, we still need update to uni_*, as it is used as tail4block process on skylake
> > https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/nodes/mkldnn_fake_quantize_node.cpp#L565
> > 
> > so like this need update:
> > https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/nodes/mkldnn_fake_quantize_node.cpp#L787
> > 
> > https://github.com/openvinotoolkit/openvino/blob/5284fb8198d781d2e4bee4d76af1d36cfee6f210/inference-engine/src/mkldnn_plugin/nodes/mkldnn_fake_quantize_node.cpp#L792
> 
> @EgorDuplensky Seems like FQ should be fixed in bounds of this PR

Those moves seems to be not considering as AVX/SSE transition, but no problem, I will correct it just in case

**Comment by EgorDuplensky** [comment]
Also the decision is to clear the rest of AVX/SSE transition issues with have found using SDE emulator.
Postponed a bit

**Comment by EgorDuplensky** [comment]
@dmitry-gorokhov @chenhu-wang 
FQ comments applied.
DeformableConvolution has also been updated with uni_v* methods

**Comment by chenhu-wang** [comment]
> @dmitry-gorokhov @chenhu-wang FQ comments applied. DeformableConvolution has also been updated with uni_v* methods

@dmitry-gorokhov @EgorDuplensky Approved assuming this PR is fix for current models, and I think we have leftover to cover in separate PR(s) for 1. check all code path that current models do not run into(some code path executed only in certain tensor shapes). 2. Make sure AVX machine use vex instruction.

As you know, the ultimate ideal goal is:
if machine is AVX or above, all use vex instructions.
if machine is below AVX, all use non-vex instructions.

So the implementation is:
For common code path(sse41/avx2/avx512), use uni_*, which check is_valid(avx) internally to use proper instructions.
For avx2 and avx512 code path, all should use vex instructions. Allow to use uni_* or explicit vex to ensure this.
For sse41 code path, considering AVX machine,  should use uni_* in this code path.
Correct me if not correct or sth missed.

**Review by chenhu-wang — APPROVED** [review]

**Review by dmitry-gorokhov — APPROVED** [review]
